### PR TITLE
Usage of Consumer API SDK for running challenges integration tests

### PR DESCRIPTION
**Pull request by Dannyps**
# Readiness checklist

-   [ ] I added/updated unit tests.
-   [x] I added/updated integration tests.
-   [x] I ensured that the PR title is good enough for the changelog.
-   [x] I labeled the PR.

### CODE REVIEW

File: BuildingBlocks/src/BuildingBlocks.SDK/Endpoints/Common/EndpointClient.cs [hunk 1]
```diff
@@ -82,9 +82,19 @@ private async Task<ApiResponse<T>> Execute<T>(HttpRequestMessage request)
             responseContent = new MemoryStream(Encoding.UTF8.GetBytes(EMPTY_RESULT));
         }
 
-        var deserializedResponseContent = JsonSerializer.Deserialize<ApiResponse<T>>(responseContent, _jsonSerializerOptions)!;
+        var deserializedResponseContent = statusCode is HttpStatusCode.OK or HttpStatusCode.Created ? JsonSerializer.Deserialize<ApiResponse<T>>(responseContent, _jsonSerializerOptions)! : new ApiResponse<T>();
```
Comment (tnotheis) [inline BuildingBlocks/src/BuildingBlocks.SDK/Endpoints/Common/EndpointClient.cs:85]: Hmm, this creates a valid ApiResponse even though the status code is an unsuccessful one, doesn't it?

File: ConsumerApi.Sdk/Client.cs [hunk 2]
```diff
@@ -18,7 +18,7 @@ namespace Backbone.ConsumerApi.Sdk;
 
 public class Client
 {
-    public Client(Configuration configuration) : this(new HttpClient { BaseAddress = UriUtils.CreateUriFromParts(configuration.BaseUrl, "api", configuration.ApiVersion) }, configuration)
+    public Client(Configuration configuration) : this(new HttpClient { BaseAddress = UriUtils.CreateUriFromParts(configuration.BaseUrl, "api", configuration.ApiVersion, "/") }, configuration)
```
Comment (tnotheis) [inline ConsumerApi.Sdk/Client.cs:21]: Instead of this "hack", I would rather suggest to update the `CreateUriFromParts` method to make sure the Uri always contains a trailing backslash.

File: ConsumerApi.Sdk/ConsumerApi.Sdk.csproj [hunk 3]
```diff
@@ -5,8 +5,13 @@
     <ImplicitUsings>enable</ImplicitUsings>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
```
Comment (tnotheis) [inline ConsumerApi.Sdk/ConsumerApi.Sdk.csproj:9]: I would prefer System.Text.Json over Newtonsoft.Json. Is there a reason why you chose Newtonsoft?

File: ConsumerApi.Sdk/Configuration.cs [hunk 4]
```diff
@@ -16,4 +16,15 @@ public class AuthenticationConfiguration
...
+    public object CloneWith(AuthenticationConfiguration newAuthentication)
+    {
+        return new Configuration()
+        {
+            Authentication = newAuthentication,
+            JsonSerializerOptions = JsonSerializerOptions,
+            BaseUrl = null!,
+            ApiVersion = null!
+        };
+    }
```
Comment (tnotheis) [inline ConsumerApi.Sdk/Configuration.cs:29]: I don't understand this method. It says "clone" but out of the 4 properties, only 1 is "cloned", while 2 are set to null (even though the properties are not even allowed to be null), and one is overridden.

File: ConsumerApi.Sdk/Endpoints/Devices/Types/SignedChallenge.cs [hunk 5]
```diff
@@ -1,7 +1,21 @@
...
+using Backbone.Crypto;
+using Newtonsoft.Json;
+
+namespace Backbone.ConsumerApi.Sdk.Endpoints.Devices.Types;
 
 public class SignedChallenge
 {
-    public required string Challenge { get; set; }
-    public required byte[] Signature { get; set; }
+    [SetsRequiredMembers]
```
Comment (tnotheis) [inline ConsumerApi.Sdk/Endpoints/Devices/Types/SignedChallenge.cs:10]: What does this attribute do?

File: ConsumerApi.Sdk/Services/AccountController.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
+﻿using Backbone.ConsumerApi.Sdk.Endpoints.Common.Crypto;
+using Backbone.ConsumerApi.Sdk.Endpoints.Devices.Types;
+using Backbone.ConsumerApi.Sdk.Endpoints.Identities.Types.Requests;
+using Backbone.Crypto;
+using Backbone.Crypto.Implementations;
+using Newtonsoft.Json;
+
+namespace Backbone.ConsumerApi.Sdk.Services;
+public class AccountController(Client client)
```
Comment (tnotheis) [inline ConsumerApi.Sdk/Services/AccountController.cs:9]: We don't use these kinds of constructors (yet)

File: ConsumerApi.Sdk/Services/AccountController.cs
Comment (tnotheis) [inline ConsumerApi.Sdk/Services/AccountController.cs:1]: Why is there an "AccountController" all of a sudden?

File: ConsumerApi.Sdk/Services/PasswordHelper.cs
Comment (tnotheis) [inline ConsumerApi.Sdk/Services/PasswordHelper.cs:1]: I don't think that we have to create random passwords, do we? Can't we just use one password for all users?

File: ConsumerApi.Tests.Integration/API/ChallengesApi.cs [hunk 7]
```diff
@@ -10,9 +10,4 @@ internal async Task<HttpResponse<Challenge>> CreateChallenge(RequestConfiguratio
     {
         return await Post<Challenge>("/Challenges", requestConfiguration);
     }
-
-    internal async Task<HttpResponse<Challenge>> GetChallengeById(RequestConfiguration requestConfiguration, string id)
-    {
-        return await Get<Challenge>($"/Challenges/{id}", requestConfiguration);
-    }
```
Comment (tnotheis) [inline ConsumerApi.Tests.Integration/API/ChallengesApi.cs:17]: Why did you remove this?

File: ConsumerApi.Tests.Integration/StepDefinitions/BaseStepDefinitions.cs [hunk 8]
```diff
@@ -29,7 +29,7 @@ public BaseStepDefinitions(IOptions<HttpConfiguration> httpConfiguration, ISigna
                 Username = "USRa",
                 Password = "a"
             },
-            ContentType = "application/json"
+            ContentType = "application/json",
```
Comment (tnotheis) [inline ConsumerApi.Tests.Integration/StepDefinitions/BaseStepDefinitions.cs:32]:
```suggestion
            ContentType = "application/json"
```

File: ConsumerApi.Tests.Integration/StepDefinitions/ChallengesApiStepDefinitions.cs
Comment (tnotheis) [inline ConsumerApi.Tests.Integration/StepDefinitions/ChallengesApiStepDefinitions.cs:1]: This seems to be solved way to complicated. But I'm not 100% certain. I will have a look by myself.

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: ConsumerApi.Sdk/ConsumerApi.Sdk.csproj (hunk 3 above)
Comment (Dannyps) [reply ConsumerApi.Sdk/ConsumerApi.Sdk.csproj]: We're using `JsonConvert.DeserializeObject` in other places, and that's available through `Newtonsoft.Json`. I'm not sure if `System.Text.Json` provides an equivalent API.

File: ConsumerApi.Tests.Integration/API/ChallengesApi.cs (hunk 7 above)
Comment (Dannyps) [reply ConsumerApi.Tests.Integration/API/ChallengesApi.cs:17]: It is no longer used. I figured we'd want to only keep SDK methods in the future.

File: ConsumerApi.Sdk/Services/PasswordHelper.cs
Comment (Dannyps) [reply ConsumerApi.Sdk/Services/PasswordHelper.cs:1]: We can. However, this is an SDK in first place, I see no harm in it providing a password helper. The fact that we use it for tests is a matter of convenience, the way I see it.

Nevertheless, this class can be removed.

File: ConsumerApi.Sdk/ConsumerApi.Sdk.csproj (hunk 3 above)
Comment (tnotheis) [reply ConsumerApi.Sdk/ConsumerApi.Sdk.csproj]: Okay, then let's keep it for now.

File: ConsumerApi.Sdk/Services/PasswordHelper.cs
Comment (tnotheis) [reply ConsumerApi.Sdk/Services/PasswordHelper.cs:1]: I don' think that I would expect the SDK to provide a password helper. You can put the class into our Tooling project, if you like.

File: ConsumerApi.Tests.Integration/API/ChallengesApi.cs (hunk 7 above)
Comment (tnotheis) [reply ConsumerApi.Tests.Integration/API/ChallengesApi.cs:17]: Okay, you're right. 👍

**Review by daniel-almeida-konkconsulting — APPROVED** [review]

File: BuildingBlocks/src/BuildingBlocks.SDK/Endpoints/Common/EndpointClient.cs [hunk 9]
```diff
@@ -108,7 +108,19 @@ private async Task<ApiResponse<T>> Execute<T>(HttpRequestMessage request)
...
+        {
+            deserializedResponseContent = JsonSerializer.Deserialize<ApiResponse<T>>(responseContent, _jsonSerializerOptions);
+        }
+        catch (JsonException)
+        {
+            deserializedResponseContent = new ApiResponse<T>
+            {
+                Result = default
+            };
+        }
```
Comment (tnotheis) [inline BuildingBlocks/src/BuildingBlocks.SDK/Endpoints/Common/EndpointClient.cs:122]: Why do you think this is a good solution?
It just covers the issue, and the user will get a seemingly valid response even though there was an error.

File: ConsumerApi.Sdk/Endpoints/Challenges/ChallengesEndpoint.cs [hunk 10]
```diff
@@ -6,8 +6,23 @@ namespace Backbone.ConsumerApi.Sdk.Endpoints.Challenges;
...
+
+    public async Task<ApiResponse<Challenge>> GetChallenge(string id)
+    {
+        return await _client.Get<Challenge>($"api/{API_VERSION}/Challenges/{id}");
+    }
+
+    public async Task<ApiResponse<Challenge>> GetChallengeUnauthenticated(string id)
+    {
+        return await _client.GetUnauthenticated<Challenge>($"api/{API_VERSION}/Challenges/{id}");
+    }
```
Comment (tnotheis) [inline ConsumerApi.Sdk/Endpoints/Challenges/ChallengesEndpoint.cs:27]: It is impossible to get a challenge without authentication

File: ConsumerApi.Tests.Integration/StepDefinitions/ChallengesApiStepDefinitions.cs [hunk 11]
```diff
@@ -1,49 +1,70 @@
...
+    private readonly ClientCredentials _clientCredentials;
 
-    public ChallengesApiStepDefinitions(IOptions<HttpConfiguration> httpConfiguration, ISignatureHelper signatureHelper, ChallengesApi challengesApi, IdentitiesApi identitiesApi, DevicesApi devicesApi) :
-        base(httpConfiguration, signatureHelper, challengesApi, identitiesApi, devicesApi)
+    public ChallengesApiStepDefinitions(IOptions<HttpConfiguration> httpConfiguration, HttpClientFactory factory)
     {
         _challengeId = string.Empty;
+        _httpClient = factory.CreateClient();
+        _clientCredentials = new ClientCredentials(httpConfiguration.Value.ClientCredentials.ClientId, httpConfiguration.Value.ClientCredentials.ClientSecret);
+        _sdk = Client.CreateUnauthenticated(_httpClient, _clientCredentials);
```
Comment (tnotheis) [inline ConsumerApi.Tests.Integration/StepDefinitions/ChallengesApiStepDefinitions.cs:29]: Is this value ever used? Not really, right? It's overriden in either `AuthenticatedUser` or `GivenTheUserIsUnauthenticated` (depending on which of those the test calls)

File: ConsumerApi.Tests.Integration/StepDefinitions/ChallengesApiStepDefinitions.cs [hunk 12]
```diff
@@ -1,49 +1,70 @@
...
+        _isAuthenticated = false;
```
Comment (tnotheis) [inline ConsumerApi.Tests.Integration/StepDefinitions/ChallengesApiStepDefinitions.cs:30]: Can be removed

File: ConsumerApi.Tests.Integration/StepDefinitions/ChallengesApiStepDefinitions.cs [hunk 13]
```diff
@@ -1,49 +1,70 @@
...
     }
 
+    [Given("the user is authenticated")]
+    public async Task AuthenticatedUser()
```
Comment (tnotheis) [inline ConsumerApi.Tests.Integration/StepDefinitions/ChallengesApiStepDefinitions.cs:34]:
```suggestion
    public async Task GivenTheUserIsAuthenticated()
```

File: ConsumerApi.Tests.Integration/api.appsettings.local.override.json
Comment (tnotheis) [inline ConsumerApi.Tests.Integration/api.appsettings.local.override.json:1]: Why don't you use the dedicated users instead of the admin ones?

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: ConsumerApi.Sdk/Endpoints/Challenges/ChallengesEndpoint.cs (hunk 10 above)
Comment (Dannyps) [reply ConsumerApi.Sdk/Endpoints/Challenges/ChallengesEndpoint.cs:27]: https://github.com/nmshd/backbone/blob/NMSHDB-83-create-identities-via-sdk/ConsumerApi.Tests.Integration/Features/Challenges/%7Bid%7D/GET.feature#L13-L17

In that case, should we ditch this scenario?

Comment (tnotheis) [reply ConsumerApi.Sdk/Endpoints/Challenges/ChallengesEndpoint.cs:27]: Yeah, I would only test scenarios that are possible with the sdk. At least for now.

File: ConsumerApi.Tests.Integration/StepDefinitions/ChallengesApiStepDefinitions.cs (hunk 11 above)
Comment (Dannyps) [reply ConsumerApi.Tests.Integration/StepDefinitions/ChallengesApiStepDefinitions.cs:29]: The only line that could be removed here is the `_sdk` attribution. Even that one is necessary if we wish to allow tests which do not start with `the user is (un)authenticated`. Should this line be removed?

Also, note that it can't really be removed. We cant set `_sdk = null!;` or we can make `_sdk` nullable.

File: ConsumerApi.Tests.Integration/api.appsettings.local.override.json
Comment (Dannyps) [reply ConsumerApi.Tests.Integration/api.appsettings.local.override.json:1]: This is in line with the admin API tests, as well as the Consumer API configuration. Should we revert this?

Comment (tnotheis) [reply ConsumerApi.Tests.Integration/api.appsettings.local.override.json:1]: I don't see a reason why we shouldn't

File: ConsumerApi.Tests.Integration/StepDefinitions/ChallengesApiStepDefinitions.cs (hunk 11 above)
Comment (tnotheis) [reply ConsumerApi.Tests.Integration/StepDefinitions/ChallengesApiStepDefinitions.cs]: > if we wish to allow tests which do not start with the user is (un)authenticated.

This will never be the case, I think.

> Should this line be removed?

Yes.

> we can make _sdk nullable.

Yes, do it.

**Review by tnotheis — APPROVED** [review]
